### PR TITLE
Initial work on grazing angle signal

### LINKED ISSUE #45: Implement calculation of vectors from grazing angle for new granular ophyd objects
As we moved to more granular ophyd objects, we need to re-implement calculation of the vectors based on the grazing angle for the supported optical elements.

See https://github.com/NSLS-II/sirepo-bluesky/blob/114384dee7b2aeb1a9635b65cc634699ef35a230/sirepo_bluesky/sirepo_bluesky.py#L133-L149 for the current implementation and https://github.com/NSLS-II/sirepo-bluesky/blob/114384dee7b2aeb1a9635b65cc634699ef35a230/sirepo_bluesky/srw_detector.py#L128-L149 where it's used.

### PR DESCRIPTION
**Pull request by AbbyGi**
Closes #45

### CODE REVIEW

File: sirepo_bluesky/sirepo_ophyd.py [hunk 1]
```diff
@@ -60,8 +60,28 @@ def trigger(self, *args, **kwargs):
...
+        self._sirepo_dict[self._sirepo_param] = value
+        self._readback = value
         return NullStatus()
 
+    def put(self, *args, **kwargs):
+        self.set(*args, **kwargs).wait()
+
+
+class GrazingAngleOrientation(Device):
+    ...
```
Comment (mrakitin) [inline sirepo_bluesky/sirepo_ophyd.py:83]: I think we can remove that unused class too.

File: sirepo_bluesky/sirepo_ophyd.py [hunk 2]
```diff
@@ -157,8 +179,9 @@ def update_components(_data):
         self._resource_document = None
         self._datum_factory = None
 
-        logger.debug(f"\nReport for {self.name}: "
-                     f"{self.connection.data['report']}\n")
+        logger.debug(
+            f"\nReport for {self.name}: " f"{self.connection.data['report']}\n"
```
Comment (mrakitin) [inline sirepo_bluesky/sirepo_ophyd.py:183]:
```suggestion
            f"\nReport for {self.name}: {self.connection.data['report']}\n"
```

File: sirepo_bluesky/sirepo_ophyd.py [hunk 3]
```diff
@@ -203,8 +226,9 @@ def trigger(self, *args, **kwargs):
         datafile = self.connection.get_datafile()
         self.report.put(json.dumps(json.loads(datafile.decode())))
 
-        logger.debug(f"\nReport for {self.name}: "
-                     f"{self.connection.data['report']}\n")
+        logger.debug(
+            f"\nReport for {self.name}: " f"{self.connection.data['report']}\n"
```
Comment (mrakitin) [inline sirepo_bluesky/sirepo_ophyd.py:230]:
```suggestion
            f"\nReport for {self.name}: {self.connection.data['report']}\n"
```

File: sirepo_bluesky/sirepo_ophyd.py [hunk 4]
```diff
@@ -260,6 +284,35 @@ def create_classes(sirepo_data, connection, create_objects=True):
...
+                            # Update vector components
+                            for cpt in [
+                                "normalVectorX",
+                                "normalVectorY",
+                                "normalVectorZ",
+                                "tangentialVectorX",
+                                "tangentialVectorY",
+                            ]:
+                                getattr(self, cpt).put(ret[cpt])
+                            return NullStatus()
```
Comment (mrakitin) [inline sirepo_bluesky/sirepo_ophyd.py:311]: Maybe we could do this operation on the original class without subclassing, such as suggested in https://stackoverflow.com/a/65571141. @AbbyGi, what do you think?

File: sirepo_bluesky/sirepo_ophyd.py [hunk 5]
```diff
@@ -60,8 +60,28 @@ def trigger(self, *args, **kwargs):
...
+        self._sirepo_dict = sirepo_dict
+        self._sirepo_param = sirepo_param
+        if sirepo_param in RESERVED_SIREPO_TO_OPHYD_ATTRS:
+            self._sirepo_param = RESERVED_SIREPO_TO_OPHYD_ATTRS[sirepo_param]
+
+    def set(self, value, *, timeout=None, settle_time=None):
+        print(f"Setting value for {self.name} to {value}")
+        self._sirepo_dict[self._sirepo_param] = value
+        self._readback = value
         return NullStatus()
```
Comment (mrakitin) [inline sirepo_bluesky/sirepo_ophyd.py:76]: That line is missing in the `DeviceWithJSONData`'s trigger method.

**Review by mrakitin — COMMENTED** [review]
Great progress! Left a few suggestions to improve the implementation even further.

**Review by jennmald — APPROVED** [review]

**Review by thomaswmorris — COMMENTED** [review]
Looks good!

File: sirepo_bluesky/sirepo_ophyd.py (hunk 4 above)
Comment (mrakitin) [reply sirepo_bluesky/sirepo_ophyd.py]: Addressed via https://github.com/NSLS-II/sirepo-bluesky/pull/46/commits/292ed50a9a98878d67942c51443cd83b69dab084.

**Review by mrakitin — APPROVED** [review]
Looks good to me! Sorry for the delay with my review.

**Comment by mrakitin** [comment]
Rebased on `master` to make sure we can cleanly merge, and all recent tests pass.

Thinking about tests, I believe a dedicated test to check the recalculation actually works would not hurt. @AbbyGi, do you mind adding a simple test for that case? Thanks!

**Review by mrakitin — APPROVED** [review]
Thanks a lot, @AbbyGi! Looks good to me.

**Pull request merged**